synapse/handlers/oidc.py

```diff
@@ -1035,6 +1139,161 @@ def _remote_id_from_userinfo(self, userinfo: UserInfo) -> str:
+
+        # Try to decode the keys in cache first, then retry by forcing the keys
+        # to be reloaded
+        jwk_set = await self.load_jwks()
+        try:
+            claims = jwt.decode(
+                logout_token,
+                key=jwk_set,
+                claims_cls=LogoutToken,
+                claims_options=claim_options,
+            )
+        except ValueError:
+            logger.info("Reloading JWKS after decode error")
+            jwk_set = await self.load_jwks(force=True)  # try reloading the jwks
+            claims = jwt.decode(
+                logout_token,
+                key=jwk_set,
+                claims_cls=LogoutToken,
+                claims_options=claim_options,
+            )
+
+        logger.debug("Decoded logout_token JWT %r; validating", claims)
+        claims.validate(leeway=120)  # allows 2 min of clock skew
+
+        # As per sec. 2.6:
+        #    4. Verify that the Logout Token contains a sub Claim, a sid Claim,
+        #       or both.
+        #    5. Verify that the Logout Token contains an events Claim whose
+        #       value is JSON object containing the member name
+        #       http://schemas.openid.net/event/backchannel-logout.
+        #    6. Verify that the Logout Token does not contain a nonce Claim.
+        # This is all verified by the LogoutToken claims class, so at this
+        # point the `sid` claim exists and is a string.
+        sid: str = claims.get("sid")
+
+        # Find in the store if we have a device associated with that SID
+        devices = await self._store.get_devices_by_auth_provider_session_id(
+            auth_provider_id=self.idp_id,
+            auth_provider_session_id=sid,
+        )
+        for device in devices:
+            user_id = device["user_id"]
+            device_id = device["device_id"]
+            logger.info(
+                "Logging out %r (device %r) via OIDC backchannel logout (sid %r).",
+                user_id,
+                device_id,
+                sid,
+            )
+            await self._device_handler.delete_device(user_id, device_id)
+
+        request.setResponseCode(200)
+        request.setHeader(b"Cache-Control", b"no-cache, no-store")
+        request.setHeader(b"Pragma", b"no-cache")
+        finish_request(request)
+
+
+class LogoutToken(JWTClaims):
+    """
+    Holds and verify claims of a logout token, as per
+    https://openid.net/specs/openid-connect-backchannel-1_0.html#LogoutToken
+    """
+
+    REGISTERED_CLAIMS = ["iss", "sub", "aud", "iat", "jti", "events", "sid"]
+
+    def validate(self, now: Optional[int] = None, leeway: int = 0) -> None:
+        """Validate everything in claims payload."""
+        super(LogoutToken, self).validate(now, leeway)
+        self.validate_sid()
+        self.validate_events()
+        self.validate_nonce()
+
+    def validate_exp(self, now: Optional[int] = None, leeway: int = 0) -> None:
```

Comment:
Do these get called automatically? (Why do we call them manually above?)

Reply:
Yep, by [`JWTClaims#validate`](https://github.com/lepture/authlib/blob/0788d705adc8232434ef497174711f2efce9879f/authlib/jose/rfc7519/claims.py#L88-L106). The manual calls are there because those are additional claims not validated by `JWTClaims`.
I'm overriding the method here to skip the regular `exp` claim validation. But now, reading the spec again, I'm not sure why past me did that, since the spec explicitly says:

>  OPs are encouraged to use short expiration times in Logout Tokens, preferably at most two minutes in the future, to prevent captured Logout Tokens from being replayable. 

So those tokens can have an `exp`, even if it's not in the [list of claims in a logout token](https://openid.net/specs/openid-connect-backchannel-1_0.html#LogoutToken)

Reply:
Removed the `exp`/`nbf` validation override